components/suggest/src/util.rs

```diff
@@ -0,0 +1,1578 @@
+/// the filter-map function.
+///
+/// The filter-map function can return `None` to halt traversal at the current
+/// step. Returning `None` sets the branching factor at that step to zero,
+/// pruning the subtree rooted at that step from the traversal space and
+/// discarding the path from the output. This is important for keeping traversal
+/// depth reasonably bounded.
+///
+/// Traversal ends and the function returns when all paths have been visited.
+/// The returned `Vec` will contain all traversal paths that weren't pruned.
+///
+/// # Examples
+///
+/// Mapping chunks in "a b c" to uppercase, up to a max chunk size of `3`:
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, _, _| {
+///     Ok(Some(vec![chunk.to_uppercase()]))
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["A", "B", "C"],
+///     vec!["A", "B C"],
+///     vec!["A B", "C"],
+///     vec!["A B C"]
+/// ]);
+/// ```
+///
+/// Same as previous but using `chunk_index` in the filter-map function to prune
+/// paths that don't start with `"a"`:
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, chunk_index, _| {
+///     if chunk_index > 0 || chunk == "a" {
+///         Ok(Some(vec![chunk.to_uppercase()]))
+///     } else {
+///         Ok(None)
+///     }
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["A", "B", "C"],
+///     vec!["A", "B C"],
+/// ]);
+/// ```
+///
+/// Same as the first example but using `path` in the filter-map function to
+/// prune paths that include "A B":
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, _, path| {
+///     if path.iter().any(|value| value == "A B") {
+///         Ok(None)
+///     } else {
+///         Ok(Some(vec![chunk.to_uppercase()]))
+///     }
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["A", "B", "C"],
+///     vec!["A", "B C"],
+///     vec!["A B C"],
+/// ]);
+/// ```
+///
+/// Mapping each chunk to multiple values:
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, _, _| {
+///     Ok(Some(vec![format!("{chunk}0"), format!("{chunk}1")]))
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["a0", "b0", "c0"],
+///     vec!["a0", "b0", "c1"],
+///     vec!["a0", "b1", "c0"],
+///     vec!["a0", "b1", "c1"],
+///     vec!["a0", "b c0"],
+///     vec!["a0", "b c1"],
+///     vec!["a1", "b0", "c0"],
+///     vec!["a1", "b0", "c1"],
+///     vec!["a1", "b1", "c0"],
+///     vec!["a1", "b1", "c1"],
+///     vec!["a1", "b c0"],
+///     vec!["a1", "b c1"],
+///     vec!["a b0", "c0"],
+///     vec!["a b0", "c1"],
+///     vec!["a b1", "c0"],
+///     vec!["a b1", "c1"],
+///     vec!["a b c0"],
+///     vec!["a b c1"]
+/// ]);
+/// ```
+pub fn filter_map_chunks<T: Clone>(
```

Comment:
I'm going to need some time to wrap my head around this one, but I have some initial thoughts

1. Can you document what "chunk" means in this context?

2. Let me propose a simpler solution, which is probably wrong, but hopefully knowing why will help me understand this better.  It seems like the requirement is to match a optional city, region, and/or weather keyword out of the input query, and the tricky part is that the keyword might be multiple-words long and you don't know which order the city/region/weather keyword will come.  Could this helper function iterate through the words and pass the string from joining the all the words that haven't been mapped to a Some value yes?  So for the "a b c" case it would:

 
- Call `mapper("a")`
  - If mapper returns `Some(_)`, then call `mapper("b")`
      - If mapper returns `Some(_)`, then call `mapper("c")`
      - If the mapper returns `None`, then call `mapper("b c")`
   - If mapper returns `None`, then call `mapper("a b")`
      - If mapper returns `Some(_)`, then call `mapper("c")`
      - If the mapper returns `None`, then call `mapper("a b c")`

For weather, the the mapper function would try to match the string to a `Token` value.

This seems simpler to understand and less complexity.  Which weather queries would it miss?

Reply:
> 1. Can you document what "chunk" means in this context?

Sure. To explain here too, it's "chunk" as in a contiguous non-overlapping sequence of items in a slice, as used in a bunch of slice and `Iterator` method names, especially https://doc.rust-lang.org/std/vec/struct.Vec.html#method.chunks

I kind of assumed "chunk" is a well known term in Rust at least w/r/t to strings and slices, but I'm not sure.

> 2. Let me propose a simpler solution

That's exactly what this function does (recursively), except it has the added complexity of allowing `mapper()` to return multiple matches and then it recurses over those matches. That's what the inner `for value in mapped_values` does. Without that for-loop, the function would behave like your example.

The reason that complexity is valuable is that a chunk can match more than one token. "New York" can be the city or the state. "Waterloo" can be Waterloo, IA or Waterloo, AL. Further, not all combinations of tokens are valid. If the query string is "waterloo al", then `[<Waterloo IA>, <AL>]` is not a valid token combination. (I'm using `<foo>` to mean a token, and `[<foo>, <bar>]` to mean a list of tokens.)

So regardless of the particular implementation, there are two steps inherent in this parsing process: (1) Map chunks to tokens, and (2) reject invalid token combinations. There's more than one way to do it, but the implementation I've chosen here is elegant IMO.

Here's an example parsing the query string "waterloo al" where the mapper only returns an `Option<Token>`, similar to your example:

1. Call `mapper("waterloo")`
2. `mapper()` returns `Some(<Waterloo IA>)`
3. Call `mapper("al")`
4. `mapper()` returns `Some(<AL>)`

We end up with tokens `[<Waterloo IA>, <AL>]`. We need a post-processing step that figures out that this combination is invalid, and then we need to start again and somehow tell the mapper to return different tokens if it can.

This example builds on that and lets the mapper return multiple tokens, but it still doesn't recurse on each mapped token:

1. Call `mapper("waterloo")`
2. `mapper()` returns `[<Waterloo IA>, <Waterloo AL>]`
3. Call `mapper("al")`
4. `mapper()` returns `[<AL>]`

We end up with `[[<Waterloo IA>, <Waterloo AL>], <AL>]`. This is better because now we can figure out valid combinations and we don't need to start again. But it still requires a post-processing step that's separate from the token parsing itself. I actually implemented this at one point, but due to this post-processing step, it was uglier than what I ended up with.

Here's what my patch does:

1. Call `mapper("waterloo", [])`
2. `mapper()` returns `Some([<Waterloo IA>, <Waterloo AL>])`
3. Call `mapper("al", [<Waterloo IA>])`
4. `mapper()` finds that "al" maps to `<AL>` but it returns `None` because `<AL>` doesn't match `<Waterloo IA>`
5. Call `mapper("al", [<Waterloo AL>])`
6. `mapper()` returns `Some([<Waterloo AL>, <AL>])`

And that's it, we're done.

Reply:
I document what "chunk" means and also added some comments to the recursive helper.